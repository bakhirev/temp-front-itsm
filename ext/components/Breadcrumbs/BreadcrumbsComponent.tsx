import styled from 'styled-components';

export const HEIGHT = 20;
export const HEIGHT_SMALL = 16;

export const BreadcrumbsComponent = styled.div<{ small?: boolean }>`
  align-items: center;
  display: flex;
  height: ${({ small }) => (small ? HEIGHT_SMALL : HEIGHT)}px;
  outline: none;
`;
