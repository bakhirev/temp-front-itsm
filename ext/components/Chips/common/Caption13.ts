import styled from 'styled-components';

import { Body2Long } from '../../Typography';

export const Caption13 = styled(Body2Long)`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;
