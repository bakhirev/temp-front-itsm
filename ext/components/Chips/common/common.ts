export type Size = 'big' | 'small';
