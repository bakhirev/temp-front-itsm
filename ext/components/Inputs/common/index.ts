export * from './interface';
export * from './hooks';
export * from './utils';
export * from './helpers';
export * from './types';
