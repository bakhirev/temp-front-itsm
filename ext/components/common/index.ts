export { DEFAULT_THEME } from './default-theme';
export { KEY_CODES } from './key-codes';
export { Z_INDEXES } from './z-indexes';
export { markerStyle } from './marker';
export { useThemeContext } from './theme-context';
