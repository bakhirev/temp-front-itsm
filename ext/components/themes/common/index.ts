export * from './box-shadow';
export * from './button';
export * from './color';
export * from './input';
export * from './interfaces';
export * from './multi-select';
export * from './tabs';
