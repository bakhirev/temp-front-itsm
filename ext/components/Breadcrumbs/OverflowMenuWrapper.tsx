import styled, { css } from 'styled-components';

import { HEIGHT as SIZE, HEIGHT_SMALL as SIZE_SMALL } from './BreadcrumbsComponent';

export const OverflowMenuWrapper = styled.div<{ small?: boolean }>`
  ${({ small }) => css`
    align-items: center;
    display: flex;
    height: ${small ? SIZE_SMALL : SIZE}px;
    justify-content: center;
    width: ${small ? SIZE_SMALL : SIZE}px;
  `};
`;
