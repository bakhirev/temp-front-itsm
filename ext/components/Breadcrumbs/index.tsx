import React, {
  createRef,
  FocusEvent,
  Fragment,
  KeyboardEvent,
  MouseEvent,
  PureComponent,
  RefObject,
} from 'react';

import { OverflowMenu } from '../OverflowMenu';
import { KEY_CODES } from '../common';

import { BreadcrumbsComponent } from './BreadcrumbsComponent';
import { OverflowMenuWrapper } from './OverflowMenuWrapper';
import { Item } from './components/Item';
import { Separator } from './components/Separator';

const MAX_WIDTH = 800;
const OVERFLOW_MENU_ID = 1;

export interface IBreadcrumbsProps {
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Список элементов */
  list: Array<IBreadcrumbsItem>;
  /** Мобильный вид компонента */
  mobile?: boolean;
  /** Маленький размер компонента */
  small?: boolean;
}

export interface IBreadcrumbsItem {
  /** Подпись к элементу */
  label: string;
  /** url элемента */
  url?: string;
}

interface IBreadcrumbsState {
  focusedItemId: number | null;
  hiddenItems: Array<IBreadcrumbsItem & { id: number }>;
  items: Array<IBreadcrumbsItem & { id: number }>;
  overflowMenuFocused: boolean;
}

export class Breadcrumbs extends PureComponent<IBreadcrumbsProps, IBreadcrumbsState> {
  selfRef: RefObject<HTMLDivElement>;
  overflowMenuRef: RefObject<HTMLDivElement>;

  constructor(props: IBreadcrumbsProps) {
    super(props);
    this.selfRef = createRef();
    this.overflowMenuRef = createRef();

    this.state = {
      focusedItemId: null,
      hiddenItems: [],
      items: [],
      overflowMenuFocused: false,
    };
  }

  static getDerivedStateFromProps({ list }: IBreadcrumbsProps) {
    return {
      items: list.map((listItem, idx) => ({
        ...listItem,
        id: idx,
      })),
    };
  }

  componentDidMount() {
    const { mobile } = this.props;
    if (mobile) return;

    const selfNode = this.selfRef.current;
    if (!selfNode) return;

    const { width } = selfNode.getBoundingClientRect();
    if (width > MAX_WIDTH) {
      this.hideItem();
    }
  }

  componentDidUpdate() {
    const { mobile } = this.props;
    if (mobile) return;

    const selfNode = this.selfRef.current;
    if (!selfNode) return;

    const { width } = selfNode.getBoundingClientRect();
    if (width > MAX_WIDTH) {
      this.hideItem();
    }
  }

  handleOverflowMenuChange = ({ value }: { value: string }) => window.location.assign(value);

  handleOverflowMenuBlur = () => {
    this.selfRef.current?.focus();
    this.setState({ overflowMenuFocused: false });
  };

  handleItemClick = (id: number) => {
    const { items } = this.state;
    const last = id === items[items.length - 1].id;

    const targetItem = this.getItemById(id);

    targetItem && targetItem.url && !last && window.location.assign(targetItem.url);
  };

  // To prevent focus event, firing before click
  handleMouseDown = (event: MouseEvent<HTMLDivElement>) => {
    this.setState({ focusedItemId: null });
    event.preventDefault();
  };

  handleFocus = ({ currentTarget, target }: FocusEvent<HTMLDivElement>) => {
    const { focusedItemId } = this.state;
    if (currentTarget === target) {
      const { items } = this.state;
      this.setState({ focusedItemId: focusedItemId ?? items[0].id });
    }
  };

  handleBlur = () => {
    const { overflowMenuFocused } = this.state;
    if (!overflowMenuFocused) {
      this.setState({ focusedItemId: null });
    }
  };

  handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
    const { focusedItemId, hiddenItems } = this.state;

    if (event.keyCode === KEY_CODES.LEFT) {
      this.focusPreviousItem();
    }

    if (event.keyCode === KEY_CODES.RIGHT) {
      this.focusNextItem();
    }

    if (event.keyCode === KEY_CODES.ENTER) {
      if (hiddenItems.length && focusedItemId === OVERFLOW_MENU_ID) {
        return;
      }
      const focusedItem = typeof focusedItemId === 'number' && this.getItemById(focusedItemId);
      focusedItem && focusedItem.url && window.location.assign(focusedItem.url);
    }
  };

  focusPreviousItem = () => {
    const { focusedItemId, hiddenItems, items } = this.state;
    if (focusedItemId === null) return;
    if (focusedItemId === OVERFLOW_MENU_ID) {
      this.overflowMenuRef.current?.blur();
    }

    const previousItemId = focusedItemId - 1;
    const firstItemId = items[0].id;

    if (!hiddenItems.length && previousItemId >= firstItemId) {
      this.setState({ focusedItemId: previousItemId });
      return;
    }

    const previousItemHidden = this.getHiddenItemById(previousItemId);

    if (previousItemHidden) {
      this.setState({ focusedItemId: OVERFLOW_MENU_ID, overflowMenuFocused: true }, () => {
        this.overflowMenuRef.current?.focus();
      });
      return;
    }

    if (previousItemId >= firstItemId) {
      this.setState({ focusedItemId: previousItemId });
    }
  };

  focusNextItem = () => {
    const { focusedItemId, hiddenItems, items } = this.state;
    if (focusedItemId === null) return;
    if (focusedItemId === OVERFLOW_MENU_ID) {
      this.overflowMenuRef.current?.blur();
    }

    const nextItemId = focusedItemId + 1;
    const lastItemId = items[items.length - 1].id;

    if (!hiddenItems.length && nextItemId < lastItemId) {
      this.setState({ focusedItemId: nextItemId });
      return;
    }

    if (nextItemId === OVERFLOW_MENU_ID && nextItemId < lastItemId) {
      this.setState({ focusedItemId: OVERFLOW_MENU_ID, overflowMenuFocused: true }, () => {
        this.overflowMenuRef.current?.focus();
      });
      return;
    }

    const nextItemHidden = this.getHiddenItemById(nextItemId);
    if (nextItemHidden) {
      const nextVisibleItem = this.getNextVisibleItem();
      const nextVisibleItemId = nextVisibleItem && nextVisibleItem.id;
      this.setState({
        focusedItemId: nextVisibleItemId || focusedItemId,
      });
      return;
    }

    if (nextItemId < lastItemId) {
      this.setState({ focusedItemId: nextItemId });
      return;
    }
  };

  getItemById = (itemId: number) => {
    const { items } = this.state;
    return items.find(({ id }) => id === itemId);
  };

  getHiddenItemById = (itemId: number) => {
    const { hiddenItems } = this.state;
    return hiddenItems.find(({ id }) => id === itemId);
  };

  getNextVisibleItem = () => {
    const { hiddenItems } = this.state;
    const lastHiddenItem = hiddenItems[hiddenItems.length - 1];
    const nextVisibleItem = this.getItemById(lastHiddenItem && lastHiddenItem.id + 1);
    return nextVisibleItem;
  };

  hideItem = () => {
    const { hiddenItems } = this.state;
    if (!hiddenItems.length) {
      const firstItemToHide = this.getItemById(OVERFLOW_MENU_ID);
      firstItemToHide &&
        hiddenItems.push(firstItemToHide) &&
        this.setState({ hiddenItems }, () => {
          this.forceUpdate();
        });
    } else {
      const lastHiddenItem = hiddenItems[hiddenItems.length - 1];
      const nextItemToHideId = lastHiddenItem.id + 1;
      const nextItemToHide = this.getItemById(nextItemToHideId);
      nextItemToHide &&
        hiddenItems.push(nextItemToHide) &&
        this.setState({ hiddenItems }, () => {
          this.forceUpdate();
        });
    }
  };

  renderContent = () => {
    const { items } = this.state;

    return items.map(({ id, label, url }) => {
      const hiddenItem = this.getHiddenItemById(id);

      if (!hiddenItem) {
        return this.renderItem({ id, label, url });
      }

      if (hiddenItem && id === OVERFLOW_MENU_ID) {
        return this.renderOverflowMenu();
      }

      return false;
    });
  };

  renderItem = ({ id, label, url }: IBreadcrumbsItem & { id: number }) => {
    const { mobile, small } = this.props;
    const { focusedItemId, items } = this.state;
    const last = id === items[items.length - 1].id;

    return (
      <Fragment key={id}>
        <Item
          focused={focusedItemId === id}
          id={id}
          label={label}
          last={last}
          mobile={mobile}
          small={small}
          url={url}
          onClick={this.handleItemClick}
        />
        {!last && <Separator small={small || mobile} />}
      </Fragment>
    );
  };

  renderOverflowMenu = () => {
    const { mobile, small } = this.props;
    const { hiddenItems } = this.state;

    const overflowMenuList = hiddenItems.map(({ id, label, url }) => ({
      id: id,
      label: label,
      value: url,
    }));

    return (
      <Fragment key="overflowMenu">
        <OverflowMenuWrapper small={small || mobile}>
          <OverflowMenu
            innerRef={this.overflowMenuRef}
            list={overflowMenuList}
            size={'small'}
            onBlur={this.handleOverflowMenuBlur}
            onChange={this.handleOverflowMenuChange}
          />
        </OverflowMenuWrapper>
        <Separator small={small || mobile} />
      </Fragment>
    );
  };

  render() {
    const { className, dataTestId, mobile, small } = this.props;
    return (
      <BreadcrumbsComponent
        className={className}
        data-test-id={dataTestId}
        ref={this.selfRef}
        small={small || mobile}
        tabIndex={1}
        onBlur={this.handleBlur}
        onFocus={this.handleFocus}
        onMouseDown={this.handleMouseDown}
        onKeyDown={this.handleKeyDown}
      >
        {this.renderContent()}
      </BreadcrumbsComponent>
    );
  }
}
