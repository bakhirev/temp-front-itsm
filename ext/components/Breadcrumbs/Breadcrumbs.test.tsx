import React, { ReactNode } from 'react';
import { mount, shallow } from 'enzyme';

import { Breadcrumbs } from '../Breadcrumbs';
import { KEY_CODES } from '../common';

jest.mock('../common', () => ({
  DEFAULT_THEME: {
    color: {
      neutral: {
        50: '',
      },
      primary: {
        60: '',
      },
    },
  },
  KEY_CODES: {
    ENTER: 13,
    LEFT: 37,
    RIGHT: 39,
  },
}));

jest.mock('../Typography', () => ({
  // eslint-disable-next-line react/display-name
  Body2Short: ({ children }: { children: ReactNode }) => <div>{children}</div>,
  // eslint-disable-next-line react/display-name
  Body1Short: ({ children }: { children: ReactNode }) => <div>{children}</div>,
}));

jest.mock('../OverflowMenu', () => {
  const OverflowMenu = () => <div />;
  return { OverflowMenu };
});

describe('Breadcrumbs', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const props = {
    list: [
      {
        label: 'Breadcrumb 1',
        url: '#1',
      },
      {
        label: 'Breadcrumb 2',
        url: '#2',
      },
      {
        label: 'Breadcrumb 3',
        url: '#3',
      },
      {
        label: 'Breadcrumb 4',
      },
    ],
  };

  it('should render component', () => {
    const wrapper = shallow(<Breadcrumbs {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with small size', () => {
    const wrapper = shallow(<Breadcrumbs {...props} small />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should open correct url when user clicks on breadcrumbs' item", () => {
    window.location.assign = jest.fn();
    const wrapper = shallow(<Breadcrumbs {...props} />);
    const item = wrapper.childAt(0).dive();
    item.simulate('click');
    expect(window.location.assign).toHaveBeenCalledWith(props.list[0].url);
  });

  it('should open correct url when user presses enter and component in focus', () => {
    window.location.assign = jest.fn();
    const wrapper = shallow(<Breadcrumbs {...props} />);
    wrapper.simulate('focus', { currentTarget: 1, target: 1 });
    wrapper.simulate('keydown', { keyCode: KEY_CODES.ENTER });

    expect(window.location.assign).toHaveBeenCalledWith(props.list[0].url);
  });

  it('should highlight next item when user presses right arrow', () => {
    window.location.assign = jest.fn();
    const wrapper = shallow(<Breadcrumbs {...props} />);
    wrapper.simulate('focus', { currentTarget: 1, target: 1 });
    wrapper.simulate('keydown', { keyCode: KEY_CODES.RIGHT });
    wrapper.simulate('keydown', { keyCode: KEY_CODES.RIGHT });
    wrapper.simulate('keydown', { keyCode: KEY_CODES.ENTER });

    expect(window.location.assign).toHaveBeenCalledWith(props.list[2].url);
  });

  it('should highlight previous item when user presses left arrow', () => {
    window.location.assign = jest.fn();
    const wrapper = shallow(<Breadcrumbs {...props} />);
    wrapper.simulate('focus', { currentTarget: 1, target: 1 });
    wrapper.simulate('keydown', { keyCode: KEY_CODES.RIGHT });
    wrapper.simulate('keydown', { keyCode: KEY_CODES.RIGHT });
    wrapper.simulate('keydown', { keyCode: KEY_CODES.LEFT });
    wrapper.simulate('keydown', { keyCode: KEY_CODES.ENTER });

    expect(window.location.assign).toHaveBeenCalledWith(props.list[1].url);
  });

  it('should render overflow menu if breadcrumbs are wider then 800px', () => {
    let widthChecks = 1;

    const wrapper = mount(<Breadcrumbs {...props} />);
    const wrapperNode = wrapper.getDOMNode();

    wrapperNode.getBoundingClientRect = (): DOMRect => {
      if (widthChecks) {
        widthChecks = 0;
        return {
          bottom: 0,
          height: 0,
          left: 0,
          right: 0,
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          toJSON: () => {},
          top: 0,
          width: 900,
          x: 0,
          y: 0,
        };
      } else {
        return {
          bottom: 0,
          height: 0,
          left: 0,
          right: 0,
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          toJSON: () => {},
          top: 0,
          width: 700,
          x: 0,
          y: 0,
        };
      }
    };

    wrapper.setProps(props);
    wrapper.update();

    expect(wrapper.find('OverflowMenu')).toHaveLength(1);
  });
});
