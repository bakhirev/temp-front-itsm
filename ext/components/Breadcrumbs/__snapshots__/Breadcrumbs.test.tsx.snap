// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs should render component 1`] = `
<styled.div
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  tabIndex={1}
>
  <Component
    focused={false}
    id={0}
    label="Breadcrumb 1"
    last={false}
    onClick={[Function]}
    url="#1"
  />
  <Component />
  <Component
    focused={false}
    id={1}
    label="Breadcrumb 2"
    last={false}
    onClick={[Function]}
    url="#2"
  />
  <Component />
  <Component
    focused={false}
    id={2}
    label="Breadcrumb 3"
    last={false}
    onClick={[Function]}
    url="#3"
  />
  <Component />
  <Component
    focused={false}
    id={3}
    label="Breadcrumb 4"
    last={true}
    onClick={[Function]}
  />
</styled.div>
`;

exports[`Breadcrumbs should render component with small size 1`] = `
<styled.div
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  small={true}
  tabIndex={1}
>
  <Component
    focused={false}
    id={0}
    label="Breadcrumb 1"
    last={false}
    onClick={[Function]}
    small={true}
    url="#1"
  />
  <Component
    small={true}
  />
  <Component
    focused={false}
    id={1}
    label="Breadcrumb 2"
    last={false}
    onClick={[Function]}
    small={true}
    url="#2"
  />
  <Component
    small={true}
  />
  <Component
    focused={false}
    id={2}
    label="Breadcrumb 3"
    last={false}
    onClick={[Function]}
    small={true}
    url="#3"
  />
  <Component
    small={true}
  />
  <Component
    focused={false}
    id={3}
    label="Breadcrumb 4"
    last={true}
    onClick={[Function]}
    small={true}
  />
</styled.div>
`;
